src/cplscheme/BaseCouplingScheme.cpp

```diff
@@ -665,22 +627,20 @@ bool BaseCouplingScheme::accelerate()
 
   // coupling iteration converged for current time window. Advance in time.
   if (convergence) {
+    moveToNextWindow();
     if (getAcceleration()) {
       getAcceleration()->iterationsConverged(getAccelerationData());
     }
```

Comment:
Is this order correct? Doesn't this mean that we now use the extrapolated data in the acceleration (e.g. to update W and V of the quasi-Newton methods)?

Reply:
After writing this example, I think the current order is incorrect. `doImplicitStep()` is quite complicated, so I will give an example to make sure I get it right:

### Some basics

* `getAccelerationData()[id].second->values()` holds the value that is exchanged. Let's say we have a Force of 5N here.
* `getAccelerationData()[id].second->previousIteration()` is a buffer. We use `storeIteration()` to write to this buffer.
* `_waveforms[id]->lastTimeWindows()` holds the value of the current window, but also of previous windows. Let's say [5N, 3N]. 3N is the Force that was computed in the last time window. We only store two samples, since we only support first order extrapolation in this example.

### Example

1. Let's assume we have just completed the last iteration in our current time window and reached convergence (just got Force 5N). So [`convergence`](https://github.com/BenjaminRodenberg/precice/blob/99ce4d134fae66a6d4aea920d5dcbe5e572e72d1/src/cplscheme/BaseCouplingScheme.cpp#L623) is true and we enter [this branch](https://github.com/BenjaminRodenberg/precice/blob/99ce4d134fae66a6d4aea920d5dcbe5e572e72d1/src/cplscheme/BaseCouplingScheme.cpp#L629-L634).
2. We now call `moveToNextWindow` and this triggers extrapolation. We use the values in `_waveforms[id]->lastTimeWindows()` ([5N, 3N]) and compute 7N through extrapolation. `moveToNextWindow` does two things:
    2.1. Set `getAccelerationData()[id].second->values()` to the newly computed value (7N)
    2.2. Update `_waveforms[id]->lastTimeWindows()` to [7N, 5N], i.e. throwing away the old value of 3N and storing the 7N.
3. After that we exit the branch and call [`storeIteration`](https://github.com/BenjaminRodenberg/precice/blob/99ce4d134fae66a6d4aea920d5dcbe5e572e72d1/src/cplscheme/BaseCouplingScheme.cpp#L643). This will save `getAccelerationData()[id].second->values()` (7N) into `getAccelerationData()[id].second->previousIteration()` (**This will be used for the convergence check in the next iteration!**).
4. We now exit `doImplicitStep()`.
5. After the participant has finished it's iteration (compute a new Force...) and called `advance()` we enter `doImplicitStep()` again. `getAccelerationData()[id].second->values()` was overwritten during the last iteration and now has the value (6N). `getAccelerationData()[id].second->previousIteration()` still holds the old value (7N).
6. We measure convergence and find out that we have not converged. Therefore, entering [this branch](https://github.com/BenjaminRodenberg/precice/blob/99ce4d134fae66a6d4aea920d5dcbe5e572e72d1/src/cplscheme/BaseCouplingScheme.cpp#L635-L640) and perform the acceleration on `getAccelerationData()`.
7. We exit the branch and continue at step 3. above.

Looking at [the version on `develop`](https://github.com/precice/precice/blob/8c2d78d03f1c9330fec41d1a8acbb33e0bcac375/src/cplscheme/BaseCouplingScheme.cpp#L658-L686) I think there is a difference: `storeData` is called **before** `extrapolateData`. Meaning: The extrapolated value is not stored and therefore also not used for convergence checks.

I will fix this now.